### ISSUE TITLE
Investigate modular architecture

### ISSUE THREAD

**Issue by aparks5** [aparks5/synthcastle#114]
Obviously without a modular GUI like VCV rack this is going to be hard but surely there is a way to describe and create new instruments or patches in text, it's just going to suck for users

lfo.out -> osc1.pitch

print command could then display all above sorted by source or sorted by destination

we could then iterate over the architecture passing each input to output for the signal flow. then per call we can poll each module

basically you need support for gates and cv

somehow each block needs to subscribe to the outputs generated by each module. 

control params need block updates, but voltage needs to be polled per sample

vcv rack uses single sample processing

how do we poll other blocks for control voltage?
perhaps its just another input and the key is the processing graph, not polling

how does modulation routing work?

**Comment by aparks5** [comment]
i had some insight about this tonight

each module generates an input that will be copied for each branch in the routing

the fundamental problem is to figure out the order of operations when modules are routed to each other

you should probably work this out by hand with a trivial patch before generalizing it

after a patch is constructed or changes, the signal flow must be recalculated

the use of mixers and splitters would be fundamental in joining multiple modules

we must also consider the use of feedback in the paths of the modules, in which case the idea of the "first module" is not totally relevant.

there is of course an initial excitation through an oscillator or the playback of a sample, but once it goes through the system for one iteration  this input could be modified

we must also calculate the total number of modules to compute a 1/N*sqrt(2) scaling factor for the total output

the signal routing can be in series or parallel as well,

although it could be said that parallel sources simply are all the independent sources fed through a mixer

the interface programatically would look something like

```
module.addModulationSource(module2, destinationParameter);
```

in C module2 might some kind of function pointer callback, but perhaps this could lead to some nasty infinite loops?

it's unclear.  perhaps this is how it should be done, where each module knows how to call other modules and use the output for specified parameters

there is also the issue of control rate and audio rate modulation

routing is easier when it is only feed forward but it limits our options

insertion of modules into the feedback path makes routing more complicated

maybe we need the concept of a patch.

we also maybe need to do some pointer magic to make this work.

if modules possess shared pointers, then perhaps any module can modify the value of the shared pointer

like

```
patch.route(lfo1,osc1.freq)
patch.route(module1.out1,module2.in2)
```

we can only route outputs to inputs
there must be some generic way of expressing an output to input relationship
what should happen when we call route?

maybe if each module maintains a pointer to the source that connects to its inputs, it can call the reference
through the pointer reference to obtain the output

the output could alert the sink when some flag `bOutputReady` is true, then when it has enough input it could compute it's own output. although all the modules could assume that the inputs are zero unless they aren't

**Comment by aparks5** [comment]
in issue #123 i refactored all processes to use a common class Module

this should help with the modular architecture. 

eventually each Module should use a common input/output interface to make chaining easier.

each Module should have a flexible number of inputs and outputs


only valid outputs should be assignable as inputs
